File: app/src/main/java/com/simplemobiletools/gallery/adapters/MyPagerAdapter.java
package com.simplemobiletools.gallery.adapters;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;

import com.simplemobiletools.gallery.Constants;
import com.simplemobiletools.gallery.activities.ViewPagerActivity;
import com.simplemobiletools.gallery.fragments.PhotoFragment;
import com.simplemobiletools.gallery.fragments.VideoFragment;
import com.simplemobiletools.gallery.fragments.ViewPagerFragment;
import com.simplemobiletools.gallery.models.Medium;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MyPagerAdapter extends FragmentStatePagerAdapter {
    private final List<Medium> mMedia;
    private final Map<Integer, ViewPagerFragment> mFragments;
    private final ViewPagerActivity mActivity;

    public MyPagerAdapter(ViewPagerActivity act, FragmentManager fm, List<Medium> media) {
        super(fm);
        mActivity = act;
        mMedia = media;
        mFragments = new HashMap<>();
    }

    @Override
    public int getCount() {
        return mMedia.size();
    }

    @Override
    public Fragment getItem(int position) {
        final Medium medium = mMedia.get(position);
        final Bundle bundle = new Bundle();
        bundle.putSerializable(Constants.MEDIUM, medium);
        ViewPagerFragment fragment;

        if (medium.getIsVideo()) {
            fragment = new VideoFragment();
        } else {
            fragment = new PhotoFragment();
        }

        mFragments.put(position, fragment);
        fragment.setArguments(bundle);
        fragment.setListener(mActivity);
        return fragment;
    }

    public void itemDragged(int pos) {
        if (mFragments.get(pos) != null) {
            mFragments.get(pos).itemDragged();
        }
    }

    public void updateUiVisibility(boolean isFullscreen, int pos) {
        for (int i = -1; i <= 1; i++) {
            final ViewPagerFragment fragment = mFragments.get(pos + i);
            if (fragment != null) {
                fragment.systemUiVisibilityChanged(isFullscreen);
            }
        }
    }

    public void updateItems(List<Medium> newPaths) {
        mMedia.clear();
        mMedia.addAll(newPaths);
        notifyDataSetChanged();
    }
}

File: app/src/main/kotlin/com/simplemobiletools/gallery/adapters/MyPagerAdapter.kt
package com.simplemobiletools.gallery.adapters

import android.os.Bundle
import android.os.Parcelable
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.PagerAdapter
import android.view.ViewGroup
import com.simplemobiletools.gallery.activities.ViewPagerActivity
import com.simplemobiletools.gallery.fragments.PhotoFragment
import com.simplemobiletools.gallery.fragments.VideoFragment
import com.simplemobiletools.gallery.fragments.ViewPagerFragment
import com.simplemobiletools.gallery.helpers.MEDIUM
import com.simplemobiletools.gallery.models.Medium

class MyPagerAdapter(val activity: ViewPagerActivity, fm: FragmentManager, val media: MutableList<Medium>) : FragmentStatePagerAdapter(fm) {
    private val fragments = HashMap<Int, ViewPagerFragment>()
    override fun getCount() = media.size

    override fun getItem(position: Int): Fragment {
        val medium = media[position]
        val bundle = Bundle()
        bundle.putSerializable(MEDIUM, medium)
        val fragment = if (medium.isVideo()) {
            VideoFragment()
        } else {
            PhotoFragment()
        }

        fragment.arguments = bundle
        fragment.listener = activity
        return fragment
    }

    override fun getItemPosition(item: Any) = PagerAdapter.POSITION_NONE

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val fragment = super.instantiateItem(container, position) as ViewPagerFragment
        fragments[position] = fragment
        return fragment
    }

    override fun destroyItem(container: ViewGroup, position: Int, any: Any) {
        fragments.remove(position)
        super.destroyItem(container, position, any)
    }

    fun getCurrentFragment(position: Int) = fragments[position]

    fun toggleFullscreen(isFullscreen: Boolean) {
        for ((pos, fragment) in fragments) {
            fragment.fullscreenToggled(isFullscreen)
        }
    }

    // try fixing TransactionTooLargeException crash on Android Nougat, tip from https://stackoverflow.com/a/43193425/1967672
    override fun saveState(): Parcelable? {
        val bundle = super.saveState() as Bundle?
        bundle?.putParcelableArray("states", null)
        return bundle
    }
}
